pennylane/ops/op_math/parametrized_hamiltonian.py

```diff
@@ -0,0 +1,172 @@
+from pennylane.operation import Operator
+from pennylane.ops.qubit.hamiltonian import Hamiltonian
+from pennylane.wires import Wires
+
+
+# pylint: disable= too-many-instance-attributes
+class ParametrizedHamiltonian:
+    r"""Callable object holding the information representing a parametrized Hamiltonian. Passing parameters to
+    the ParametrizedHamiltonian returns an Operator representing the Hamiltonian for that set of parameters.
+
+    The Hamiltonian can be represented as a linear combination of other operators, e.g.,
+    :math:`H(v, t) = H_\text{drift} + \sum_j f_j(v, t) H_j`, where the :math:`v` are trainable parameters,
+    and t is time.
+
+    For example, a time-dependent ``ParametrizedHamiltonian`` with a single trainable parameter, :math:`a`, could be :math:`H = 2 * X_1 X_2 + sin(a, t) * Y_1 Y_2`
+
+    Args:
+        coeffs (Union[float, callable]): coefficients of the Hamiltonian expression, which may be constants or
+            parametrized functions. All functions passed as ``coeffs`` must accept the same parameters as arguments.
+        observables (Iterable[Observable]): observables in the Hamiltonian expression, of same length as coeffs
+
+    A ParametrizedHamiltonian is callable, and passing parameters to the ParametrizedHamiltonian will return an
+    Operator representing an instance of the Hamiltonian with the specified parameter values.
+
+    **Example:**
+
+    A ParametrizedHamiltonian can be created by passing a list of coefficients (scalars or functions), as well as
+    a list of corresponding observables. The functions must have identical signatures, though they may not all
+    use all the parameters.
+
+    >>> def f1(params, t): return np.sin(params[0]*t)
+    >>> def f2(params, t): return params[1] * np.cos(t)
+
+    The functions, along with scalar coefficients, can then be used to initialize a ParametrizedHamiltonian,
+    which will be split into a fixed and parametrized term. The fixed term is an Operator, while the parametrized
+    term must be initialized with concrete parameters to obtain an Operator.
+    >>> coeffs = [2, f1, f2]
+    >>> obs = [qml.PauliX(0)@qml.PauliX(1), qml.PauliY(0)@qml.PauliY(1), qml.PauliZ(0)@qml.PauliZ(1)]
+    >>> H = ParametrizedHamiltonian(coeffs, obs)
+    >>> H.H_fixed
+      2*(PauliX(wires=[0]) @ PauliX(wires=[1]))
+    >>> H.H_parametrized([2.5, 3.6], t)
+      (0.5984721441039564*(PauliY(wires=[0]) @ PauliY(wires=[1]))) + (1.9450883011253033*(PauliZ(wires=[0]) @ PauliZ(wires=[1])))
+
+    The resulting object can be passed parameters, and will return an Operator representing the
+    ParametrizedHamiltonian with the specified parameters:
+
+    >>> H([1.2, 2.3], 0.5)
+    (2*(PauliX(wires=[0]) @ PauliX(wires=[1]))) + ((0.5646424733950354*(PauliY(wires=[0]) @ PauliY(wires=[1]))) + (2.0184398923478573*(PauliZ(wires=[0]) @ PauliZ(wires=[1]))))
+
+    """
+
+    def __init__(self, coeffs, observables):
+
+        if qml.math.shape(coeffs)[0] != len(observables):
```

Comment:
Careful, `qml.math.shape` will convert `coeffs` into an array, and this might be problematic when `coeffs` is a list of arrays with different shapes (given that the new `numpy` version will raise an error).

Better to use `if len(coeffs) != len(observables)` here.

Reply:
Thanks!